#52 - Refactor unit test layout and naming to be more BDD orientated

AwaNoodle opened the pull request:
Converted PercentageFeatureSwitchTests to class / sub class style
Split out compound tests to use test data 
Factored out repeated code

AwaNoodle committed:
Factored out recurring code to create a feature switch

AwaNoodle committed:
Factored out recurring code to create a feature switch

AwaNoodle committed:
IsOnIffRandomNumberBelowThreshold changed to use TestCase data rather than multiple cases in one test

AwaNoodle committed:
IsOnIffRandomNumberBelowThreshold changed to use TestCase data rather than multiple cases in one test

AwaNoodle committed:
Changed names to fit with more BDD style
Split out UsesValueFromCookieWhenCookieExists to be seperate tests

AwaNoodle committed:
Changed names to fit with more BDD style
Split out UsesValueFromCookieWhenCookieExists to be seperate tests

AwaNoodle committed:
Switched to using the NUnit Result field in the TestCase attribute

AwaNoodle committed:
Switched to using the NUnit Result field in the TestCase attribute

lholman commented on Femah.Core.Tests/PercentageFeatureSwitchTests.cs:
```diff
@@ -7,111 +6,87 @@
 
 namespace Femah.Core.Tests
 {
+    using System.Runtime.Remoting.Messaging;
```
Not required.

lholman commented:
Hey Mark, thanks.  However could you make define TestCaseData as you did here please https://github.com/lloydstone/femah/blob/master/Femah.Core.Tests/InvalidRequestTestData.cs

The resulting tests as is aren't particularly nice to read

![femahpullrequest80-unittestdataneedsdefining](https://cloud.githubusercontent.com/assets/362197/2632432/602e95c4-be64-11e3-9350-07864455e54d.PNG)

lholman committed:
Merge branch '48-include-psake-in-femah-build'

AwaNoodle commented:
@lholman Aye, I'll get that done now.

AwaNoodle committed:
Removed unused remoting include

AwaNoodle committed:
Set test case names for cookie tests

AwaNoodle committed:
Added test case names for percentage threshold tests

AwaNoodle commented:
Changed them to to read like this now. I've used the pre-existing comments for the random number threshold test for the test case names. 

![image](https://cloud.githubusercontent.com/assets/3598109/2660835/9d6f3516-c025-11e3-9d74-e30a7f503199.png)

AwaNoodle committed:
Merge branch 'MovingPercentageFeatureSwitchTests' of https://github.com/AwaNoodle/femah into MovingPercentageFeatureSwitchTests

Conflicts:
	Femah.Core.Tests/PercentageFeatureSwitchTests.cs

AwaNoodle commented:
I've rebased onto the main and added in the changes. The merge is a bit puzzling as it thinks my remote was out of step with my local even though they were at the same stage :/

lholman commented:
@AwaNoodle Neat, being an absolute pain but can we be consistent, either spaces or underscores in the test case names?  
I notice we used underscores before, could you update the `ReturnsFalse_IfRandomNumberIsBelowThreshold` ones to use underscores please, then we're good to go I think.

AwaNoodle committed:
Fixed naming issue with threshold tests to be consitent with other test names

AwaNoodle committed:
Added missing Is to threshold test

AwaNoodle commented:
![image](https://cloud.githubusercontent.com/assets/3598109/2674328/b66061dc-c0ff-11e3-9ed4-dd2b0bce01e4.png)

ta-da!

lholman commented on Femah.Core.Tests/PercentageFeatureSwitchTests.cs:
```diff
@@ -9,109 +8,83 @@ namespace Femah.Core.Tests
+            [TestCase(0.4, 30, Result = false, TestName = "ReturnsFalse_IfRandomNumberIsAboveThreshold")]
+            [TestCase(0.1, 10, Result = false, TestName = "ReturnsFalse_IfRandomNumberIsSameAsThreshold")]
+            [TestCase(0.0, 0, Result = false, TestName = "ReturnsFalse_IfRandomNumberIsSameAsThreshold_AndThresholdIsZero")]
+            public bool ReturnsFalse_IfRandomNumberIsBelowThreshold(double randomNumber, int percentageOn)
```
Can we rename this to something more appropriate such as? 'ReturnsSwitchState_DependantOnComparisonOfRandomNumberToThreshold'

lholman commented on Femah.Core.Tests/PercentageFeatureSwitchTests.cs:
```diff
@@ -9,109 +8,83 @@ namespace Femah.Core.Tests
+            [Test]
+            [TestCase(true, Result = true, TestName="ReturnsTrueFromCookie_IfCookieExistsWithTrueValue")]
+            [TestCase(false, Result = false, TestName="ReturnsFalseFromCookie_IfCookieExistsWithFalseValue")]
+            public bool ReturnsValueFromCookie_IfCookieExists(bool cookieValue)
```
I'm confused what these tests are actually doing, I realise we've taken it verbatim @AwaNoodle . I think it may be that we're using 'True' and 'False' as example strings that's throwing me, why do we need to test twice that we can return something in a cookie, or am I missing something? @hotstone Any thoughts?

Oh I see, I don't think I understood what the value of the cookie was doing, ignore the above.

It needs renaming though, maybe 'ReturnsSwitchState_DependantOnCookieValue' ?

AwaNoodle commented:
@lholman That makes more sense to me. Update coming soon.

AwaNoodle committed:
Updated test names to reflect comments

AwaNoodle committed:
Fixed naming mistakes in test cases

lholman commented:
Boom, this _is_ merged, I think I made a mistake with the local merge (it got a bit hairy with the rebase) and I used a local branch so I believe github didn't spot the association to this PR, anyway, thanks fella, keep it coming.
